File: gradle-plugin/src/main/java/org/komapper/gradle/KomapperPlugin.java
package org.komapper.gradle;

import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.tasks.TaskProvider;
import org.komapper.gradle.codegen.GenerateTask;

public class KomapperPlugin implements Plugin<Project> {
  @Override
  public void apply(Project project) {
    KomapperExtension extension =
        project.getExtensions().create(KomapperExtension.NAME, KomapperExtension.class, project);
    registerGenerateTasks(project, extension);
  }

  private void registerGenerateTasks(Project project, KomapperExtension extension) {
    String description = "Generate entities and definitions";
    String prefix = KomapperExtension.NAME;
    String suffix = "Generator";
    TaskProvider<?> aggregateTask =
        project
            .getTasks()
            .register(
                prefix + suffix,
                it -> {
                  it.setDescription(description + ".");
                  it.setGroup(KomapperExtension.NAME);
                  it.getOutputs().upToDateWhen(__ -> false);
                });
    extension
        .getGenerators()
        .all(
            generator -> {
              String name = prefix + capitalize(generator.getName()) + suffix;
              TaskProvider<GenerateTask> task =
                  project.getTasks().register(name, GenerateTask.class, generator);
              task.configure(
                  it -> {
                    it.setDescription(description + " for " + generator.getName() + ".");
                    it.setGroup(KomapperExtension.NAME);
                    it.getOutputs().upToDateWhen(__ -> false);
                  });
              aggregateTask.configure(it -> it.dependsOn(task));
            });
  }

  private String capitalize(String text) {
    return Character.toUpperCase(text.charAt(0)) + text.substring(1);
  }
}

File: gradle-plugin/src/main/kotlin/org/komapper/gradle/KomapperPlugin.kt
package org.komapper.gradle

import org.gradle.api.Plugin
import org.gradle.api.Project
import org.komapper.gradle.codegen.GenerateTask

@Suppress("unused")
class KomapperPlugin : Plugin<Project> {

    override fun apply(project: Project) {
        val extension =
            project.extensions.create(KomapperExtension.NAME, KomapperExtension::class.java, project)
        registerGenerateTasks(project, extension)
    }

    private fun registerGenerateTasks(project: Project, extension: KomapperExtension) {
        val description = "Generate entities and definitions"
        val prefix = KomapperExtension.NAME
        val suffix = "Generator"
        val aggregateTask = project.tasks.register(prefix + suffix) {
            it.description = "$description."
            it.group = KomapperExtension.NAME
            it.outputs.upToDateWhen { false }
        }
        extension.generators.all { generator ->
            val task = project.tasks
                .register("$prefix${generator.name.capitalize()}$suffix", GenerateTask::class.java, generator)
            task.configure {
                it.description = "$description for ${generator.name}."
                it.group = KomapperExtension.NAME
                it.outputs.upToDateWhen { false }
            }
            aggregateTask.configure { it.dependsOn(task) }
        }
    }
}
